### PR TITLE
Add baseline check for triangulation

### PR DESCRIPTION
**Pull request by ayushbaid**

### CODE REVIEW

File: gtsfm/data_association/point3d_initializer.py [hunk 1]
```diff
@@ -200,6 +199,35 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
 
         return track_3d, avg_track_reproj_error, is_cheirality_failure
 
+    def is_small_triangulation_angle(
+        self, point_3d: np.ndarray, camera_indices: List[int], angle_threshold_deg: float = 1.5
+    ) -> bool:
+        """Check if all pairs of cameras have a small triangulation angle.
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:205]: thanks for the PR, Ayush. Maybe you could add a URL ref here to where other libraries also do this?

File: gtsfm/utils/triangulation.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+Authors: Ayush Baid
+"""
+import numpy as np
+from gtsam import PinholeCameraCal3Bundler
+
+
+def calculate_triangulation_angle_in_degrees(
+    camera_1: PinholeCameraCal3Bundler, camera_2: PinholeCameraCal3Bundler, point_3d: np.ndarray
+) -> float:
+    """Calculates the angle formed at the 3D point by the rays backprojected from 2 cameras.
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:12]: looks good. would a bit of ascii art help make the triangle relationship even clearer?

File: gtsfm/utils/triangulation.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    Args:
+        camera_1: the first camera.
+        camera_2: the second camera.
+        point_3d: the 3d point to compute the angle at.
+
+    Returns:
+        the angle formed at the 3d point, in radians.
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:20]: degrees, not radians?

File: gtsfm/utils/triangulation.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+def calculate_triangulation_angle_in_degrees(
+    camera_1: PinholeCameraCal3Bundler, camera_2: PinholeCameraCal3Bundler, point_3d: np.ndarray
+) -> float:
+    """Calculates the angle formed at the 3D point by the rays backprojected from 2 cameras.
+
+    Args:
+        camera_1: the first camera.
+        camera_2: the second camera.
+        point_3d: the 3d point to compute the angle at.
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:17]: "to compute the angle at" is a bit vague. maybe we can add a bit more detail

File: gtsfm/data_association/point3d_initializer.py [hunk 5]
```diff
@@ -176,15 +174,16 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
-                rank_tol=SVD_DLT_RANK_TOL,
-                optimize=True,
+                camera_track, measurement_track, rank_tol=SVD_DLT_RANK_TOL, optimize=True,
             )
         except RuntimeError:
             is_cheirality_failure = True
             return None, None, is_cheirality_failure
 
+        # check for small baseline case
+        if self.is_small_triangulation_angle(triangulated_pt, inlier_idxs):
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:184]: Ayush, could you show the full metrics on master vs. on this branch? and a qualitative image before and after?

**Comment by ayushbaid** [comment]
I tried experimenting with two different thresholds for the baseline angle (1.5 degrees from Colmap and 4 degrees from Theia).

[2021-07-06 22:43:08,884 INFO bundle_adjustment.py line 119 16911] initial error: 35362920.83
[2021-07-06 22:43:08,884 INFO bundle_adjustment.py line 120 16911] final error: 90805.44
[2021-07-06 22:43:09,289 INFO bundle_adjustment.py line 127 16911] [Result] Number of tracks before filtering: 5288
[2021-07-06 22:43:09,910 INFO bundle_adjustment.py line 135 16911] [Result] Number of trac

| Metric | No filter | 1.5 deg | 4 deg | 
| -------- | ---------- | ---------- | -------- |
| % of tracks discarded due to small baseline | N/A | 0.6% | 21.5% |
| [BA] [pre-filtering] mean reproj error | 0.451 | 452 | 0.491 |
| [BA] [pre-filtering] median reproj error | 0.236 | 0.236 | 0.257 |
| [BA] [pre-filtering] number of tracks | 7168 | 7115 | 5288 |
| [BA] [post-filtering] number of tracks | 7072 | 7019 | 5193 |
| [DA] mean track length | 4.51 | 4.53 | 5.40 |
| [DA] median track length | 4.0 | 4.0 | 5.0 |


## No filtering  
### BA input
3d
### BA output
3d

## 1.5 deg filter
### BA input
3d
### BA output
3d


## 4 deg filter
### BA input
3d
### BA output
3d

File: gtsfm/data_association/point3d_initializer.py (hunk 5 above)
Comment (ayushbaid) [reply gtsfm/data_association/point3d_initializer.py:184]: I have added the metrics on the PR. Also, by default this check will remain off.

File: gtsfm/data_association/point3d_initializer.py (hunk 1 above)
Comment (ayushbaid) [reply gtsfm/data_association/point3d_initializer.py:205]: added them now

File: gtsfm/utils/triangulation.py (hunk 3 above)
Comment (ayushbaid) [reply gtsfm/utils/triangulation.py:20]: corrected. thanks

File: gtsfm/utils/triangulation.py (hunk 4 above)
Comment (ayushbaid) [reply gtsfm/utils/triangulation.py:17]: does this look better?

File: gtsfm/data_association/point3d_initializer.py [hunk 6]
```diff
@@ -192,13 +205,46 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
+
+
+        Args:
+            point_3d: the 3d point which has been triangulated.
+            camera_indices: indices of the cameras associated with the triangulation.
+
+        Returns:
+            True if all the pairs of camera subtend a small angle, and hence resulting in a small baseline track.
+        """
+
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:231]: thanks for the updates, Ayush. nit: prefer no extra whitespace here.

File: gtsfm/data_association/point3d_initializer.py [hunk 7]
```diff
@@ -192,13 +205,46 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
+        for i1, i2 in itertools.combinations(camera_indices, 2):
+            camera_i1 = self.track_camera_dict[i1]
+            camera_i2 = self.track_camera_dict[i2]
+
+            triangulation_angle = triangulation_utils.calculate_triangulation_angle_in_degrees(
+                camera_i1, camera_i2, point_3d
+            )
+
+            if triangulation_angle > self.min_baseline_thresh:
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:243]: so our logic here is that if any pair has a low triangulation error, then we discard the track.

but doesnt COLMAP instead say that if any camera pair has a large enough triangulation angle, then the whole track is kept?

i would probably prefer COLMAP's metric.

File: gtsfm/data_association/point3d_initializer.py [hunk 8]
```diff
@@ -192,13 +205,46 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
             track_3d.add_measurement(i, uv)
 
-        return track_3d, avg_track_reproj_error, is_cheirality_failure
+        return track_3d, avg_track_reproj_error, TriangulationResult.SUCCESS
+
+    def is_small_triangulation_angle(self, point_3d: np.ndarray, camera_indices: List[int]) -> bool:
+        """Check if all pairs of cameras have a small triangulation angle.
+
+        References:
+        - https://github.com/colmap/colmap/blob/513fef3e39f99b50279bf5aee1c5b597cba53c3b/src/base/reconstruction.cc#L616
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:220]: maybe we can add a link to the implementation of the COLMAP fn here, also: https://github.com/colmap/colmap/blob/dev/src/base/reconstruction.cc#L1412

Comment (johnwlambert) [reply gtsfm/data_association/point3d_initializer.py:220]: Like `"Based off of the implementation here: https://github.com/colmap/colmap/blob/dev/src/base/reconstruction.cc#L1412"`

File: gtsfm/utils/triangulation.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/utils/triangulation.py:12]: thanks for the ascii art. much clearer!

File: gtsfm/utils/triangulation.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    Returns:
+        the angle formed at the 3d point, in degrees.
+    """
+    camera_center_1: np.ndarray = camera_1.pose().translation()
+    camera_center_2: np.ndarray = camera_2.pose().translation()
+
+    # compute the three squared edge lengths of the triangle formed between camera centers and the 3d point
+    def squared_dist_fn(x, y):
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:32]: good to add type hints even for callbacks

File: gtsfm/utils/triangulation.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
+"""Utilities to support triangulation.
+
+Authors: Ayush Baid
+"""
+import numpy as np
+from gtsam import PinholeCameraCal3Bundler
+
+
+def calculate_triangulation_angle_in_degrees(
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:9]: can we cite COLMAP's equivalent fn in the docstring? https://github.com/colmap/colmap/blob/dev/src/base/triangulation.cc#L122

File: gtsfm/utils/triangulation.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+    # compute the three squared edge lengths of the triangle formed between camera centers and the 3d point
+    def squared_dist_fn(x, y):
+        return np.sum(np.square(x - y), axis=None)
+
+    baseline_squared = squared_dist_fn(camera_center_1, camera_center_2)
+    ray_length_1_squared = squared_dist_fn(camera_center_1, point_3d)
+    ray_length_2_squared = squared_dist_fn(camera_center_2, point_3d)
+
+    # use the law of cosines to estimate the angle at the 3d point
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:39]: i think the colmap comment is a bit clearer:
```// Using "law of cosines" to compute the enclosing angle between rays.```

File: gtsfm/utils/triangulation.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        the angle formed at the 3d point, in degrees.
+    """
+    camera_center_1: np.ndarray = camera_1.pose().translation()
+    camera_center_2: np.ndarray = camera_2.pose().translation()
+
+    # compute the three squared edge lengths of the triangle formed between camera centers and the 3d point
+    def squared_dist_fn(x, y):
+        return np.sum(np.square(x - y), axis=None)
+
+    baseline_squared = squared_dist_fn(camera_center_1, camera_center_2)
```
Comment (johnwlambert) [inline gtsfm/utils/triangulation.py:35]: isn't it simpler to just create two vectors (from point to each camera), and compute the angle between them using the typical dot product cosine equation?

a \dot b = |a| |b| cos \theta

We already have that function implemented here: https://github.com/borglab/gtsfm/blob/master/gtsfm/utils/geometry_comparisons.py#L281

File: gtsfm/data_association/point3d_initializer.py (hunk 7 above)
Comment (ayushbaid) [reply gtsfm/data_association/point3d_initializer.py:243]: My logic is the same too. if I have an angle which is greater than the min threshold, I return false for the *small triangulation angle* check. So I am essentially doing the same thing.

File: gtsfm/data_association/point3d_initializer.py (hunk 8 above)
Comment (ayushbaid) [reply gtsfm/data_association/point3d_initializer.py:220]: Sure

File: gtsfm/utils/triangulation.py (hunk 9 above)
Comment (ayushbaid) [reply gtsfm/utils/triangulation.py:32]: Thanks

File: gtsfm/utils/triangulation.py (hunk 12 above)
Comment (ayushbaid) [reply gtsfm/utils/triangulation.py:35]: yep makes sense. Thanks!

**Comment by ayushbaid** [comment]
On the skydio 8 image dataset with the SIFT Frontend (using E matrix in computation):



Metric | No filter  | 4 deg
-- | -- | -- | --
[BA] mean reproj error | 1.627  | 1.485
[BA] [pre-filtering] number of tracks | 1224 |  539
[BA] [post-filtering] number of tracks | 153 |  214

Images:
no baseline check:
8crane_nobaselinecheck

4 degree baseline check:
8crane_4degbaselinecheck

File: gtsfm/data_association/data_assoc.py [hunk 13]
```diff
@@ -39,12 +36,16 @@ class DataAssociation(NamedTuple):
         min_track_len: min length required for valid feature track / min nb of supporting views required for a landmark
                        to be valid.
         mode: triangulation mode, which dictates whether or not to use robust estimation.
+        min_angle_thresh (optional): threshold of the angle at the 3D point by backprojected rays from 2 cameras for
```
Comment (johnwlambert) [inline gtsfm/data_association/data_assoc.py:39]: thanks for the updates, Ayush. I prefer something along the lines of `min_tri_angle_deg` like COLMAP.

How about: 
```python
min_tri_angle_deg (optional): minimum allowed angle in degrees between any two rays
     during triangulation.  Each ray originates at a camera and shoots towards the triangulated 3d point.
     Small angles indicate small baselines, which are unreliable. Defaults to None (check inactive).
```

File: gtsfm/data_association/point3d_initializer.py [hunk 14]
```diff
@@ -51,12 +60,15 @@ class Point3dInitializer(NamedTuple):
         track_cameras: Dict of cameras and their indices.
         mode: triangulation mode, which dictates whether or not to use robust estimation.
         reproj_error_thresh: threshold on reproj errors for inliers.
+        min_angle_thresh (optional): threshold of the angle at the 3D point by backprojected rays from 2 cameras for
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:63]: same, i think i prefer `min_tri_angle_deg`, and the docstring i wrote above

File: gtsfm/data_association/point3d_initializer.py [hunk 15]
```diff
@@ -192,13 +205,46 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
+            return None, avg_track_reproj_error, TriangulationResult.HIGH_REPROJECTION_ERROR
 
         track_3d = SfmTrack(triangulated_pt)
         for i, uv in inlier_track.measurements:
             track_3d.add_measurement(i, uv)
 
-        return track_3d, avg_track_reproj_error, is_cheirality_failure
+        return track_3d, avg_track_reproj_error, TriangulationResult.SUCCESS
+
+    def is_small_triangulation_angle(self, point_3d: np.ndarray, camera_indices: List[int]) -> bool:
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:216]: Would it be clearer to say
```"""Checks if no camera pairs within a track have a large enough baseline to reliably triangulate.```

File: gtsfm/data_association/point3d_initializer.py [hunk 16]
```diff
@@ -192,13 +205,46 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
+        """Check if all pairs of cameras have a small triangulation angle.
+
+        References:
+        - https://github.com/colmap/colmap/blob/513fef3e39f99b50279bf5aee1c5b597cba53c3b/src/base/reconstruction.cc#L616
+        - https://github.com/sweeneychris/TheiaSfM/blob/d2112f15aa69a53dda68fe6c4bd4b0f1e2fe915b/src/theia/sfm/set_outlier_tracks_to_unestimated.cc#L121 # noqa: E501
+        - https://github.com/colmap/colmap/blob/dev/src/base/reconstruction.cc#L1412"
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:222]: nit: trailing quotation mark

File: gtsfm/data_association/point3d_initializer.py [hunk 17]
```diff
@@ -192,13 +205,46 @@ def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optiona
...
 
-        return track_3d, avg_track_reproj_error, is_cheirality_failure
+        return track_3d, avg_track_reproj_error, TriangulationResult.SUCCESS
+
+    def is_small_triangulation_angle(self, point_3d: np.ndarray, camera_indices: List[int]) -> bool:
+        """Check if all pairs of cameras have a small triangulation angle.
+
+        References:
+        - https://github.com/colmap/colmap/blob/513fef3e39f99b50279bf5aee1c5b597cba53c3b/src/base/reconstruction.cc#L616
+        - https://github.com/sweeneychris/TheiaSfM/blob/d2112f15aa69a53dda68fe6c4bd4b0f1e2fe915b/src/theia/sfm/set_outlier_tracks_to_unestimated.cc#L121 # noqa: E501
```
Comment (johnwlambert) [inline gtsfm/data_association/point3d_initializer.py:221]: nit: prefer the URL without the commit hash
```
https://github.com/sweeneychris/TheiaSfM/blob/master/src/theia/sfm/set_outlier_tracks_to_unestimated.cc#L121
```

File: gtsfm/data_association/data_assoc.py (hunk 13 above)
Comment (johnwlambert) [reply gtsfm/data_association/data_assoc.py:39]: Theia also uses something along these lines: `min_triangulation_angle_degrees`

File: tests/utils/test_triangulation.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+class TestTriangulationUtils(unittest.TestCase):
+    def test_calculate_triangulation_angle_in_degrees(self):
+        """Test the computation of triangulation angle using a simple example."""
+
+        camera_center_1 = np.array([0, 0, 0])
+        camera_center_2 = np.array([10, 0, 0])
+        point_3d = np.array([5, 0, 10])
+
+        expected = 53.1301024
```
Comment (johnwlambert) [inline tests/utils/test_triangulation.py:21]: 53 degrees is a bit confusing, could we do a simpler one we can do in our head, like 45 deg?

**Comment by ayushbaid** [comment]
Update: Does not improve metrics right now. So closing the PR.

**Comment by johnwlambert** [comment]
Sounds good. Thanks for checking on this!

**Pull request closed without merging**